allennlp/common/file_utils.py

```diff
@@ -0,0 +1,72 @@
+import os
+import base64
+import logging
+from urllib.parse import urlparse
+
+import requests
+import tqdm
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+CACHE_ROOT = os.path.expanduser(os.path.join('~', '.allennlp'))
+DATASET_CACHE = os.path.join(CACHE_ROOT, "datasets")
+
+def url_to_filename(url: str) -> str:
+    """
+    Converts a url into a filename in a reversible way.
+    """
+    url_bytes = url.encode('utf-8')
+    b64_bytes = base64.b64encode(url_bytes)
+    return b64_bytes.decode('utf-8')
+
+def filename_to_url(filename: str) -> str:
+    """
+    Recovers the the url from the encoded filename.
+    """
+    filename_bytes = filename.encode('utf-8')
+    url_bytes = base64.b64decode(filename_bytes)
+    return url_bytes.decode('utf-8')
+
+def cached_path(url_or_filename: str, cache_dir: str = DATASET_CACHE) -> str:
+    """
+    Given something that might be a URL (or might be a local path),
+    determine which. If it's a URL, download the file and cache it, and
+    return the path to the cached file. If it's already a local path,
+    make sure the file exists and then return the path.
+    """
+    parsed = urlparse(url_or_filename)
+
+    if parsed.scheme in ('http', 'https'):
+        # URL, so get it from the cache (downloading if necessary)
+        return get_from_cache(url_or_filename, cache_dir)
+    elif parsed.scheme == '' and os.path.exists(url_or_filename):
+        # File, and it exists.
+        return url_or_filename
+    elif parsed.scheme == '':
+        # File, but it doesn't exist.
+        raise FileNotFoundError("file {} not found".format(url_or_filename))
+    else:
```

Comment:
Is it worth checking for things like `file://`?  (I'd say probably not, just checking.)